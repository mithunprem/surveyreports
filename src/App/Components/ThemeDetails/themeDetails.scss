.theme-details {
  .card .card-body {
    padding: 5px;
  }

  &.margin-green {
    border-color: green;
  }

  &.margin-red {
    border-color: red;
  }

  &.margin-yellow {
    border-color: #f5b238d4;
  }
}
