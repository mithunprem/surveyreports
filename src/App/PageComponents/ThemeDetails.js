import React, { Component, Fragment } from 'react';
import PropTypes from 'prop-types';
import {
  Col, Button, Card, Collapse,
  CardHeader, CardFooter, CardBody
} from 'reactstrap';
import QuestionsTable from './QuestionsTable';
import colourCodesForRating from '../Utils/ColourCodesForRating';


export default class ThemeDetails extends Component {

  state = {
    showQuestions: false,
    themeRating: 0,
    outlineColor: ""
  }

  componentDidMount() {
    const { theme } = this.props;
    const themeRating = theme.averageRating.toFixed(2);
    let outlineColor = colourCodesForRating(themeRating);

    this.setState({ themeRating, outlineColor });
  }

  toggleQuestionView = () => {
    this.setState({
      showQuestions: !this.state.showQuestions
    })
  }

  render() {
    const { theme } = this.props;
    if (!theme) return null;

    const { showQuestions, themeRating, outlineColor } = this.state;
    const buttonText =
      showQuestions ? 'Hide responses' : 'View responses in detail';

    return (
      <Fragment>
        <Col xs={12} sm={12} md={6} xl={6} className="mb-4">
          <Card body outline className={`theme-details margin-${outlineColor}`}>
            <CardHeader tag="h5">{theme.name}</CardHeader>
            <CardBody>
              <Button size="sm" color="link" onClick={this.toggleQuestionView}>
                {buttonText}
              </Button>
              <Collapse isOpen={showQuestions}>
                <QuestionsTable questions={theme.questions} />
              </Collapse>
            </CardBody>
            <CardFooter className="text-muted">
              Average rating : {( themeRating )}
            </CardFooter>
          </Card>
        </Col>
      </Fragment>
    );
  }
}

ThemeDetails.propTypes = {
  theme: PropTypes.object
}
