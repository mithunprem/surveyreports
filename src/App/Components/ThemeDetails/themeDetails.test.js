import React from 'react';
import { create } from 'react-test-renderer';
import ThemeDetails from '../../Components/ThemeDetails';

describe("ThemeDetails", () => {

  const theme = {
    name: 'The Work',
    questions: [
      { description: 'Test question 1',
        survey_responses: [
          {id: 1, question_id: 4, respondent_id: 1, response_content: "4"},
          {id: 2, question_id: 4, respondent_id: 2, response_content: "4"}
        ],
        averageRating: 5 }
    ],
    averageRating: 4.5
  };


  test("should render without throwing an error", () => {
    const themeDetails = create(<ThemeDetails theme={theme} />).toJSON();
    expect(themeDetails).toMatchSnapshot();
  });

  test("should render the theme card with appropriate color coding", () => {
    const themeDetails = create(<ThemeDetails theme={theme} />).toJSON();
    const className = themeDetails.children[0].props.className;
    // The theme has a rating of 4.5 and hence this it should have
    // a margin-green class associated with it.
    expect(className).toBe('theme-details margin-green card card-body');
  });

});
