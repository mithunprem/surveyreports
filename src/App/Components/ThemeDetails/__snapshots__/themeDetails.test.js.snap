// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThemeDetails should render without throwing an error 1`] = `
<div
  className="mb-4 col-12 col-sm-12 col-md-6 col-xl-6"
>
  <div
    className="theme-details margin-green card card-body"
  >
    <h5
      className="card-header"
    >
      The Work
    </h5>
    <div
      className="card-body"
    >
      <button
        aria-label={null}
        className="btn btn-link btn-sm"
        onClick={[Function]}
        type="button"
      >
        View responses in detail
      </button>
      <div
        className="collapse"
        style={Object {}}
      >
        <table
          className="questions-table table table-striped"
        >
          <thead>
            <tr>
              <th>
                Question
              </th>
              <th
                className="average-rating-header"
              >
                Average Rating
              </th>
            </tr>
          </thead>
          <tbody>
            <tr>
              <td>
                Test question 1
              </td>
              <td
                className="average-rating-value text-green"
              >
                5.00
              </td>
            </tr>
          </tbody>
        </table>
      </div>
    </div>
    <div
      className="text-muted card-footer"
    >
      Average rating : 
      4.50
    </div>
  </div>
</div>
`;
